src/vim9instr.c

```diff
@@ -2401,10 +2401,15 @@ generate_store_var(
 		{
 		    isntype_T   isn_type = ISN_STORES;
 
+		    // If sn_import_autoload, generate ISN_STOREEXPORT (instead
+		    // of ISN_STORES) if there is no sn_autoload_export or if
+		    // the name does not start with "s:". Note that ISN_STORES
+		    // assumes/requires the name to start with "s:".
 		    if (SCRIPT_ID_VALID(scriptvar_sid)
 			     && SCRIPT_ITEM(scriptvar_sid)->sn_import_autoload
-			     && SCRIPT_ITEM(scriptvar_sid)->sn_autoload_prefix
-								       == NULL)
+			     && ((SCRIPT_ITEM(scriptvar_sid)
+						  ->sn_autoload_prefix == NULL)
+				|| STRCMP("s:", name) != 0))
```

Comment:
The `s:` prefix for script-local variable is not supported for script-local variables in Vim9 scripts.
Under what condition the name will start with `s:` in a Vim9 script?

Reply:
Looking at #14606 pure vim9script is generating a `ISN_STORES` which it should not do. But as to

> Under what condition the name will start with s: in a Vim9 script

I have virtually no experience with legacy script, and even less with using legacy and vim9script together but consider this legacy script 
```
let s:foo = 'init'
def F()
    s:foo = ':def assign to script local'
enddef
call F()
disa F
echo ' '
echo 'after F():' s:foo
```
which outputs/generates the ISN_STORES which is causing the trouble in pure vim9script
```
F
    s:foo = ':def assign to script local'
   0 PUSHS ":def assign to script local"
   1 STORES s:foo in /tmp/legacy.vim
   2 RETURN void

after F(): :def assign to script local
```
And look a few lines down in this function where it says
```
// "s:" may be included in the name.
```
Look in the original code in vim9execute
```
// store s: variable in old script or autoload import
case ISN_STORES:
case ISN_STOREEXPORT:
    {
        int		sid = iptr->isn_arg.loadstore.ls_sid;
        hashtab_T	*ht = &SCRIPT_VARS(sid);
        char_u	*name = iptr->isn_arg.loadstore.ls_name;
        dictitem_T	*di = find_var_in_ht(ht, 0,
                                iptr->isn_type == ISN_STORES
                                         ? name + 2 : name, TRUE);
```
When it sees the `ISN_STORES` instruction it skips the assumed `s:`, using `name + 2` for the `ht` lookup, but name was `val` in the test that causes the problem, so it looks up `l`.

Reply:
Thanks for the explanation.